core/src/main/scala/cats/data/Prod.scala

```diff
@@ -0,0 +1,108 @@
+package cats
+package data
+
+/**
+ * [[Prod]] is a product to two independent functor values.
+ *
+ * See: [[https://www.cs.ox.ac.uk/jeremy.gibbons/publications/iterator.pdf The Essence of the Iterator Pattern]]
+ */
+sealed trait Prod[F[_], G[_], A] {
+  def first: F[A]
+  def second: G[A]
+}
+object Prod extends ProdInstances {
+  def apply[F[_], G[_], A](first0: => F[A], second0: => G[A]): Prod[F, G, A] = new Prod[F, G, A] {
```

Comment:
I am actually unsure if I should attempt laziness at all, since most of the other parts in Cats that are involved, like the `CompositeApplicative` in https://github.com/non/cats/blob/master/core/src/main/scala/cats/Applicative.scala.

The potential usefulness in keeping lazy/by-name is computing `first` and `second` in parallel as a future extension. I haven't figured out if that's essentially the same as making a composite with `Future[_]`.